pbft/pbft.cpp

```diff
@@ -352,21 +352,35 @@ pbft::forward_request_to_primary(const bzn_envelope& request_env)
     this->failure_detector->request_seen(req_hash);
 }
 
+
 void
-pbft::maybe_record_request(const pbft_msg& msg, const std::shared_ptr<pbft_operation>& op)
+pbft::maybe_record_request(const pbft_msg& msg, const bzn_envelope& env, const std::shared_ptr<pbft_operation>& op)
 {
-    if (msg.has_request() && !op->has_request())
+    if (!op->has_request())
     {
-        if (this->crypto->hash(msg.request()) != msg.request_hash())
+        if (msg.request_type() == PBFT_REQUEST_PAYLOAD)
         {
-            LOG(info) << "Not recording request because its hash does not match";
-            return;
+            assert(env.piggybacked_requests_size()); // we must have at least 1 request
+            if (this->crypto->hash(env.piggybacked_requests(0)) != msg.request_hash())
+            {
+                LOG(info) << "Not recording request because hashes do not match";
+                return;
+            }
+            op->record_request(env.piggybacked_requests(0));
+        }
+        else
+        {
+            if (this->crypto->hash(msg.request()) != msg.request_hash())
+            {
+                LOG(info) << "Not recording request because hashes do not match";
+                return;
+            }
+            op->record_request(msg.request());
         }
-
-        op->record_request(msg.request());
     }
```

Comment:
Upon further reading, it seems to me as if this method is used in two ways
- saving the request that is part of some pbft message corresponding to an internal request
- saving the (first only) request piggybacked to a preprepare corresponding to a payload request

And that further these cases are invoked from different contexts and share little if any code. Should they not just be different methods, then?